shared-bindings/bitmapfilter/__init__.c

```diff
@@ -0,0 +1,432 @@
+    return args[ARG_bitmap].u_obj;
+}
+
+MP_DEFINE_CONST_FUN_OBJ_KW(bitmapfilter_lookup_obj, 0, bitmapfilter_lookup);
+
+//| def false_color(
+//|     bitmap: displayio.Bitmap,
+//|     palette: displayio.Palette,
+//|     mask: displayio.Bitmap | None,
+//| ) -> displayio.Bitmap:
+//|     """Convert the image to false color using the given palette
+//|
+//|     The ``bitmap``, which must be in RGB565_SWAPPED format, is converted into false color.
+//|
+//|     The ``palette``, which must be of length 256, is used as a look-up table.
+//|
+//|     Each pixel is converted to a luminance (brightness/greyscale) value
+//|     in the range 0..255, then the corresponding palette entry is looked up and
+//|     stored in the bitmap.
+//|
+//|     ``mask`` is another image to use as a pixel level mask for the operation.
+//|     The mask should be an image the same size as the image being operated on.
+//|     Only pixels set to a non-zero value in the mask are modified.
+//|     """
+//|
+STATIC mp_obj_t bitmapfilter_false_color(size_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_args) {
+    enum { ARG_bitmap, ARG_palette, ARG_mask };
+    static const mp_arg_t allowed_args[] = {
+        { MP_QSTR_bitmap, MP_ARG_REQUIRED | MP_ARG_OBJ, { .u_obj = MP_OBJ_NULL } },
+        { MP_QSTR_palette, MP_ARG_REQUIRED | MP_ARG_OBJ, { .u_obj = MP_OBJ_NULL } },
+        { MP_QSTR_mask, MP_ARG_OBJ, { .u_obj = MP_ROM_NONE } },
+    };
+    mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
+    mp_arg_parse_all(n_args, pos_args, kw_args, MP_ARRAY_SIZE(allowed_args), allowed_args, args);
+
+    mp_arg_validate_type(args[ARG_bitmap].u_obj, &displayio_bitmap_type, MP_QSTR_bitmap);
+    displayio_bitmap_t *bitmap = MP_OBJ_TO_PTR(args[ARG_bitmap].u_obj);
+
+    mp_arg_validate_type(args[ARG_palette].u_obj, &displayio_palette_type, MP_QSTR_palette);
+    displayio_palette_t *palette = MP_OBJ_TO_PTR(args[ARG_palette].u_obj);
+    mp_arg_validate_length(palette->color_count, 256, MP_QSTR_palette);
+
+    displayio_bitmap_t *mask = NULL;
+    if (args[ARG_mask].u_obj != mp_const_none) {
+        mp_arg_validate_type(args[ARG_mask].u_obj, &displayio_bitmap_type, MP_QSTR_mask);
+        mask = MP_OBJ_TO_PTR(args[ARG_mask].u_obj);
+    }
+
+    shared_module_bitmapfilter_false_color(bitmap, mask, palette->colors);
+    return args[ARG_bitmap].u_obj;
+}
+
+MP_DEFINE_CONST_FUN_OBJ_KW(bitmapfilter_false_color_obj, 0, bitmapfilter_false_color);
+STATIC const mp_rom_map_elem_t bitmapfilter_module_globals_table[] = {
+    { MP_ROM_QSTR(MP_QSTR___name__), MP_ROM_QSTR(MP_QSTR_bitmapfilter) },
+    { MP_ROM_QSTR(MP_QSTR_morph), MP_ROM_PTR(&bitmapfilter_morph_obj) },
+    { MP_ROM_QSTR(MP_QSTR_mix), MP_ROM_PTR(&bitmapfilter_mix_obj) },
+    { MP_ROM_QSTR(MP_QSTR_solarize), MP_ROM_PTR(&bitmapfilter_solarize_obj) },
+    { MP_ROM_QSTR(MP_QSTR_false_color), MP_ROM_PTR(&bitmapfilter_false_color_obj) },
+    { MP_ROM_QSTR(MP_QSTR_lookup), MP_ROM_PTR(&bitmapfilter_lookup_obj) },
```

Comment:
Are these names consistent with something like imagemagick? Maybe example image magick commands should be put in the docs. Inline images would be awesome too.

Reply:
Sadly, not really consistent with anything, but nothing is consistent with anything else anyway.

I can rename anything in particular if you have a specific request.

 * "solarize": this is the standard name for this algorithm
 * "morph": this is the name of the routine in openmv. the parameters are compatible or almost-compatible. imagemagick calls this "convert -morphology" so at least it's partially the same (there's also "convert -morph" which is NOT the same, that's some kind of image blending algorithm). PIL calls this "kernel".
 * "mix": this is my name. imagemagic calls it "convert -color-matrix" which I kinda like as a name. Confusingly, their documentation calls for 5x5 or 6x6 matrices (!). PIL hides this in "convert", where it shows an example of converting from the XYZ colorspace to RGB.
 * "false_color": openmv calls this "to_ironbow" and has a palette argument which defaults to the "ironbow palette". imagemagic calls this "clut". It looks like PIL also hides this in "convert" when converting from a "L" format to a "P" format.
 * "lookup": openmv does not have this function, as far as I found. imagemagic doesn't have this function, as far as I found. PIL has "LUT3D" which is a much more powerful version of this (which needs a prohibitively large table). All these tools have different image operations that can be implemented via lookup table, such as gamma correction & histogram equalization.

Reply:
This info would be really useful to have in the function docs. :-)